src/utils/SpokeUtils.ts

```diff
@@ -0,0 +1,142 @@
+ * @param maxSearches The maximum number of searches to perform. This is used to prevent infinite loops.
+ * @returns The block range that contains the deposit ID.
+ * @note  // We want to find the block range that satisfies these conditions:
+ *        // - the low block has deposit count <= targetDepositId
+ *        // - the high block has a deposit count > targetDepositId.
+ *        // This way the caller can search for a FundsDeposited event between [low, high] that will always
+ *        // contain the event emitted when deposit ID was incremented to targetDepositId + 1. This is the same transaction
+ *        // where the deposit with deposit ID = targetDepositId was created.
+ */
+export async function getBlockRangeForDepositId(
+  targetDepositId: number,
+  initLow: number,
+  initHigh: number,
+  maxSearches: number,
+  spokePool: SpokePool,
+  deploymentBlock = 0
+): Promise<{
+  low: number;
+  high: number;
+}> {
+  // Get the most recent block from the spoke pool contract.
+  const mostRecentBlockNumber = await spokePool.provider.getBlockNumber();
+  // Set the initial high block to the most recent block number or the initial high block, whichever is smaller.
+  initHigh = Math.min(initHigh, mostRecentBlockNumber);
+
+  // Define a mapping of block numbers to deposit IDs. This is used to cache the deposit ID at a block number
+  // so we don't need to make an eth_call request to get the deposit ID at a block number more than once.
+  const queriedIds: Record<number, number> = {};
+
+  // Define a llambda function to get the deposit ID at a block number. This function will first check the
+  // queriedIds cache to see if the deposit ID at the block number has already been queried. If not, it will
+  // make an eth_call request to get the deposit ID at the block number. It will then cache the deposit ID
+  // in the queriedIds cache.
+  const _getDepositIdAtBlock = async (blockNumber: number): Promise<number> => {
+    if (queriedIds[blockNumber] === undefined) {
+      const resultId = await getDepositIdAtBlock(spokePool, blockNumber);
+      if (!Number.isInteger(resultId)) {
+        throw new Error("Invalid deposit count");
+      }
+      queriedIds[blockNumber] = resultId;
+    }
+    return queriedIds[blockNumber];
+  };
+
+  // Sanity check to ensure that init Low is greater than or equal to zero.
+  if (initLow < deploymentBlock) {
+    throw new Error("Binary search failed because low must be >= 0");
+  }
+
+  // Sanity check to ensure that initHigh is greater than or equal to initLow.
+  if (initLow > initHigh) {
+    throw new Error("Binary search failed because low > high");
+  }
+
+  // Sanity check to ensure that maxSearches is greater than zero.
+  if (maxSearches <= 0) {
+    throw new Error("maxSearches must be > 0");
+  }
+
+  // Sanity check to ensure that deploymentBlock is greater than or equal to zero.
+  if (deploymentBlock < 0) {
+    throw new Error("deploymentBlock must be >= 0");
+  }
```

Comment:
Basic input validation takes up a lot of space in this function, and it's a series of checks that we probably never expect to trigger. Thoughts on squashing it?

```suggestion
  const errors: [boolean, string][] = [ 
    [initLow < deploymentBlock, "Binary search failed because low must be >= 0"],
    [initLow > initHigh, "Binary search failed because low > high"],
    [maxSearches <= 0, "maxSearches must be > 0"],
    [deploymentBlock < 0, "deploymentBlock must be >= 0"],
  ];  

  errors.forEach(([condition, error]) => {
    if (!condition) {
      throw new Error(error);
    }   
  });
```

Reply:
I'd still argue that these specific checks are just simple and basic input validation...I don't see any risk of them hiding any complexity or creating confusion. The error messages that are thrown are also matched almost 1:1 by the comments, so I don't really see how this kind of redundancy helps the reader. To me, it just seems like this validation just occupies vertical real estate and draws attention away from where the real complexity is (which is now very clear and well-documented, btw).

Reply:
Fair enough - I've modified the tests here: https://github.com/across-protocol/sdk-v2/pull/383/commits/0134433543e75a83862c2abedc2084585a7d0ec4